File: app/src/main/java/com/funworld/example/heogiutien/Expense.java
package com.funworld.example.heogiutien;

import android.os.Parcel;
import android.os.Parcelable;

import com.activeandroid.annotation.Column;
import com.activeandroid.annotation.Table;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

/**
 * Created by ThanhTD on 6/25/2017.
 */

@Table(name = "Expense")
public class Expense implements Parcelable{
    @Column(name = "expense_id", unique = true)
    private int expenseId;
    @Column(name = "type")
    private String type;
    @Column (name = "related_person")
    private String relatedPerson;
    @Column(name = "money_amount")
    private int amount;
    @Column(name = "reason")
    private String reason;
    @Column(name = "time")
    private Date time;
    private String relatedExpense; // cac chi tieu lien quan, tach chi tieu...

    public Expense(int expenseId, String type, String relatedPerson, int amount, String reason, Date time, String relatedExpense) {
        this.expenseId = expenseId;
        this.type = type;
        this.relatedPerson = relatedPerson;
        this.amount = amount;
        this.reason = reason;
        this.time = time;
        this.relatedExpense = relatedExpense;
    }

    protected Expense(Parcel in) {
        expenseId = in.readInt();
        type = in.readString();
        relatedPerson = in.readString();
        amount = in.readInt();
        reason = in.readString();
        relatedExpense = in.readString();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(expenseId);
        dest.writeString(type);
        dest.writeString(relatedPerson);
        dest.writeInt(amount);
        dest.writeString(reason);
        dest.writeString(relatedExpense);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<Expense> CREATOR = new Creator<Expense>() {
        @Override
        public Expense createFromParcel(Parcel in) {
            return new Expense(in);
        }

        @Override
        public Expense[] newArray(int size) {
            return new Expense[size];
        }
    };

    public int getExpenseId() {
        return expenseId;
    }

    public void setExpenseId(int expenseId) {
        this.expenseId = expenseId;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getRelatedPerson() {
        return relatedPerson;
    }

    public void setRelatedPerson(String relatedPerson) {
        this.relatedPerson = relatedPerson;
    }

    public int getAmount() {
        return amount;
    }

    public void setAmount(int amount) {
        this.amount = amount;
    }

    public String getReason() {
        return reason;
    }

    public void setReason(String reason) {
        this.reason = reason;
    }

    public Date getTime() {
        return time;
    }

    public void setTime(Date time) {
        this.time = time;
    }

    public String getRelatedExpense() {
        return relatedExpense;
    }

    public void setRelatedExpense(String relatedExpense) {
        this.relatedExpense = relatedExpense;
    }


    public Date setDateFromString(String date) throws ParseException {
        SimpleDateFormat sf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        sf.setLenient(true);
        return sf.parse(date);
    }
}

File: app/src/main/java/com/funworld/heogiutien/data/model/Expense.kt
package com.funworld.heogiutien.data.model

import android.os.Parcel
import android.os.Parcelable
import android.support.annotation.StringDef
import com.activeandroid.Model
import com.activeandroid.annotation.Column
import com.activeandroid.annotation.Table
import com.activeandroid.query.Select
import org.joda.time.DateTime
import org.joda.time.format.DateTimeFormat


/**
 * Created by ThanhTD on 6/25/2017.
 */

@Table(name = "Expense")
class Expense() : Model(), Parcelable {

//    @Column(name = "expense_id", unique = true, onUniqueConflict = Column.ConflictAction.REPLACE)
//    var expenseId: Int = 0
    //loai chi tieu: chi (-), them vao(+)
    @Column(name = "type")
    @ExpenseType lateinit var type: String
    //luong tien dung cho chi tieu
    @Column(name = "money_amount")
    var amount: Int = 0
    //muc dich cua chi tieu
    @Column(name = "purpose")
    lateinit var purpose: String
    // nguon tien dung cho chi tieu
    @Column(name = "resource", onUpdate = Column.ForeignKeyAction.CASCADE, onDelete = Column.ForeignKeyAction.CASCADE)
    lateinit var resourceId: Resource
    @Column(name = "note")
    var note: String? = null
    //thoi gian tao chi tieu
    @Column(name = "created_at")
    var createdAt: Long = 0
    //thoi gian cua lan chinh sua cuoi cung
    @Column(name = "updated_at")
    var updatedAt: Long = 0

    var relatedExpense: String = "" // cac chi tieu lien quan, tach chi tieu...

    constructor(parcel: Parcel) : this() {
        type = parcel.readString()
        amount = parcel.readInt()
        purpose = parcel.readString()
        resourceId = parcel.readParcelable(Resource::class.java.classLoader)
        note = parcel.readString()
        createdAt = parcel.readLong()
        updatedAt = parcel.readLong()
        relatedExpense = parcel.readString()
    }

    fun getCreatedTime()= DateTimeFormat.forPattern("HH:mm").print(createdAt)

    fun getCreatedDateTime() = DateTimeFormat.forPattern("YYYY-MM-dd HH:mm").print(createdAt)

    override fun toString() = "[type: $type, amount: $amount, purpose: $purpose, createdAt: $createdAt, resource id: $resourceId]"

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(type)
        parcel.writeInt(amount)
        parcel.writeString(purpose)
        parcel.writeParcelable(resourceId, flags)
        parcel.writeString(note)
        parcel.writeLong(createdAt)
        parcel.writeLong(updatedAt)
        parcel.writeString(relatedExpense)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Expense> {
        override fun createFromParcel(parcel: Parcel): Expense {
            return Expense(parcel)
        }

        override fun newArray(size: Int): Array<Expense?> {
            return arrayOfNulls(size)
        }

        const val DEPOSIT_TYPE = "deposit"
        const val WITHDRAW_TYPE = "withdraw"

        @StringDef(DEPOSIT_TYPE, WITHDRAW_TYPE)
        @Retention(AnnotationRetention.SOURCE)
        annotation class ExpenseType

        fun get50LatestExpenses(resource: Resource): MutableList<Expense> {
            return Select()
                    .from(Expense::class.java)
                    .where("resource = ?", resource.id)
                    .orderBy("created_at ASC")
                    .limit(50)
                    .execute()
        }

        fun getInPeriod(resource: Resource, from: Long, to: Long): MutableList<Expense> {
            return Select()
                    .from(Expense::class.java)
                    .where("resource = ? AND created_at > ? AND created_at < ?", resource.id, from, to)
                    .orderBy("created_at ASC")
                    .execute()
        }

        fun getByDate(resource: Resource, date: Long): MutableList<Expense> {
            return getInPeriod(resource,
                    date,
                    DateTime(date).withTime(23, 59, 59, 999).millis)
        }

        fun getSumOfPeriod(resource: Resource, from: Long, to: Long): Int{
//            return getByDate(resource, date).map { it.amount }.sum()
            return getInPeriod(resource, from, to).sumBy { it.amount }
        }

        fun getAll() = Select().from(Expense::class.java).execute<Expense>()

        fun getByResource(resource: Resource) = Select().from(Expense::class.java)
                .where("resource=?", resource.id).execute<Expense>()
    }
}
